pkg/ebpf/c/tracee.bpf.c

```diff
@@ -4455,6 +4506,83 @@ int BPF_KPROBE(trace_security_inode_mknod)
     return events_perf_submit(&data, SECURITY_INODE_MKNOD, 0);
 }
 
+SEC("kprobe/do_splice")
+TRACE_ENT_FUNC(do_splice, DIRTY_PIPE_SPLICE);
+
+SEC("kretprobe/do_splice")
+int BPF_KPROBE(trace_ret_do_splice)
+{
+    args_t saved_args;
+    if (load_args(&saved_args, DIRTY_PIPE_SPLICE) != 0) {
+        // missed entry or not traced
+        return 0;
+    }
+
+    event_data_t data = {};
+    if (!init_event_data(&data, ctx))
+        return 0;
+
+    // dirty_pipe_splice is a splice to a pipe which results that the last page copied could be modified (the
+    // PIPE_BUF_CAN_MERGE flag is on in the pipe_buffer struct).
+    if (!event_chosen(DIRTY_PIPE_SPLICE))
+        return 0;
+
+    struct file *out_file = (struct file*)saved_args.args[2];
+    struct inode *out_inode = READ_KERN(out_file->f_inode);
+    struct pipe_inode_info *out_pipe = READ_KERN(out_inode->i_pipe);
+    // Check that output is a pipe
+    if (!out_pipe) {
+        return 0;
+    }
+
+    struct pipe_buffer *last_write_page_buffer = get_last_write_pipe_buffer(out_pipe);
+    unsigned int out_pipe_last_buffer_flags = READ_KERN(last_write_page_buffer->flags);
+    // This is not a dirty-pipe splice if CAN_MERGE flag is off.
+    if ((out_pipe_last_buffer_flags & PIPE_BUF_FLAG_CAN_MERGE) == 0) {
+        return 0;
+    }
+
+    u64 out_inode_number = READ_KERN(out_inode->i_ino);
+    unsigned int out_pipe_last_buffer_next_data_offset = READ_KERN(last_write_page_buffer->offset) + READ_KERN(last_write_page_buffer->len);
+
+    struct file *in_file = (struct file*)saved_args.args[0];
+    struct inode *in_inode = READ_KERN(in_file->f_inode);
+    u64 in_inode_number = READ_KERN(in_inode->i_ino);
+    unsigned short in_file_type = READ_KERN(in_inode->i_mode) & S_IFMT;
+    void *in_file_path = get_path_str(GET_FIELD_ADDR(in_file->f_path));
+    size_t write_len = (size_t)saved_args.args[4];
+
+    loff_t *off_in_addr = (loff_t *)saved_args.args[1];
+    // In kernel v5.10 the pointer passed was no longer of the user, so flexibility is needed to read it
+    loff_t off_in;
+    #ifndef CORE
+    #if LINUX_VERSION_CODE < KERNEL_VERSION(5, 10, 0)
+    off_in = READ_USER(*off_in_addr);
+    #else
+    off_in = READ_KERN(*off_in_addr);
+    #endif
+    #else // CORE
+    struct public_key_signature *check;
```

Comment:
ok, this DEFINITELY needs more explanation (saying its a mitigation for checking kernel versions based on diff types) =D or else unadvertised readers might be totally crazy =D.